pkg/storageincentives/redistributionstate.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storageincentives
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/settlement/swap/erc20"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/storageincentives/redistribution"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"math/big"
+	"sync"
+)
+
+const redistributionStatusKey = "redistribution_state"
+const loggerNameNode = "nodestatus"
+
+type RedistributionState struct {
+	stateStore     storage.StateStorer
+	erc20Service   erc20.Service
+	logger         log.Logger
+	overlay        swarm.Address
+	mtx            sync.Mutex
+	status         Status
+	currentBalance *big.Int
+	contract       redistribution.Contract
+}
+
+// Status provide internal status of the nodes in the redistribution game
+type Status struct {
+	Phase           PhaseType
+	IsFrozen        bool
+	Round           uint64
+	LastWonRound    uint64
+	LastPlayedRound uint64
+	LastFrozenRound uint64
+	Block           uint64
+	Reward          *big.Int
+	Fees            *big.Int
+}
+
+func NewRedistributionState(logger log.Logger, stateStore storage.StateStorer, erc20Service erc20.Service, contract redistribution.Contract) RedistributionState {
```

Comment:
`RedistributionState` methods like `SetCurrentEvent`, `IsFrozen`, `SetLastFrozenRound` and others all seem to do `saveStatus()` at the end. This operation will potentially access disk therefore could take significant time in some situations. 

Would it be possible to decouple disk access from logic of these methods?

-----

As easiest approach would it make sense to call `saveStatus` in goroutine? Let's assume `IsFronzen` method (but this would apply to other methods as well)

```
func (n *RedistributionState) IsFrozen(f bool) {
	n.mtx.Lock()
	defer n.mtx.Unlock()
	n.status.IsFrozen = f
	go saveStatus(n.stateStore, n.status) // 1*
}
```
1* `saveStatus` should changed to accept copy of `status` variable and store it in background; thus `IsFrozen` would decouple from disk io

I am not able to estimate how often is `saveStatus` called but would it make sense to reduce this method calls? For example would it be possible to `saveStatus` only once every second if there are some changes to the state already made?